File: DataBindingStudy/app/src/main/java/com/jie/databinding/BaseObservableActivity.java
package com.jie.databinding;

import androidx.appcompat.app.AppCompatActivity;
import androidx.databinding.DataBindingUtil;
import androidx.databinding.Observable;

import com.jie.databinding.databinding.ActivityBaseobservableBinding;
import com.jie.databinding.model.Goods;

import android.os.Bundle;
import android.util.Log;


import java.util.Random;

public class BaseObservableActivity extends AppCompatActivity {
    private final String TAG = "SecondActivity";
    private Goods mGoods;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        ActivityBaseobservableBinding baseobservableBinding = DataBindingUtil.setContentView(this, R.layout.activity_baseobservable);
        mGoods = new Goods("dogecoin", "buy somethings", 0.25F);
        baseobservableBinding.setGoods(mGoods);
        baseobservableBinding.setGoodsHandler(new GoodsHandler());
        mGoods.addOnPropertyChangedCallback(new Observable.OnPropertyChangedCallback() {
            @Override
            public void onPropertyChanged(Observable sender, int propertyId) {
                if (propertyId == BR.name) {
                    Log.d(TAG, "onPropertyChanged: wyj goods name changes");
                } else if (propertyId == BR.details) {
                    Log.d(TAG, "onPropertyChanged: wyj goods details changed.");
                } else if (propertyId == BR._all) {
                    Log.d(TAG, "onPropertyChanged: wyj goods all property changed.");
                } else {
                    Log.d(TAG, "onPropertyChanged: wyj unknow.");
                }
            }
        });
    }

    public class GoodsHandler {
        public void changeGoodsName() {
            mGoods.setName("dogeCoin super" + new Random().nextInt(100));
            mGoods.setPrice(new Random().nextFloat());
        }

        public void changeGoodDetails() {
            mGoods.setDetails("in 2022, tesla goods can be buy with dogeCoin, said Elon Musk."+ new Random().nextInt(100));
            mGoods.setPrice(new Random().nextFloat());
        }
    }
}

File: DataBindingStudy/app/src/main/java/com/jie/databinding/ui/activity/BaseObservableActivity.kt
package com.jie.databinding.ui.activity

import android.os.Bundle
import android.util.Log
import androidx.appcompat.app.AppCompatActivity
import androidx.databinding.DataBindingUtil
import androidx.databinding.Observable
import com.jie.databinding.BR
import com.jie.databinding.R
import com.jie.databinding.databinding.ActivityBaseobservableBinding
import com.jie.databinding.model.GoodsEntity
import java.util.*

class BaseObservableActivity : AppCompatActivity() {
    companion object {
        private const val TAG = "BaseObservableActivity"
    }

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        val baseobservableBinding = DataBindingUtil.setContentView<ActivityBaseobservableBinding>(
            this,
            R.layout.activity_baseobservable
        )
        val goods: GoodsEntity = GoodsEntity()
        goods.name = "Doge Coin"
        goods.detail = "Buy something."
        goods.price = 0.25f
        baseobservableBinding.goods = goods
        baseobservableBinding.goodsHandler = GoodsHandler(goods)
        title = "BaseObservable Study"
        goods.addOnPropertyChangedCallback (object : Observable.OnPropertyChangedCallback() {
            override fun onPropertyChanged(sender: Observable?, propertyId: Int) {
                when(propertyId) {
                    BR.name -> Log.d(TAG, "onPropertyChanged: wyj goods name changed")
                    BR.detail -> Log.d(TAG, "onPropertyChanged: wyj goods detail changed")
                    BR._all -> Log.d(TAG, "onPropertyChanged: wyj goods all property changed")
                    else -> Log.d(TAG, "onPropertyChanged: wyj unknow.")
                }
            }
        })
    }

    inner class GoodsHandler(val goods: GoodsEntity) {
        fun changeGoodsName() {
            goods.name = "dogeCoin super ${Random().nextInt(100)}"
            goods.price = Random().nextFloat()
        }

        fun changeGoodsDetails() {
            goods.detail = "in 2022, tesla goods can be buy with dogeCoin, said Elon Musk. ${Random().nextInt(100)}"
            goods.price = Random().nextFloat()
        }
    }
}